### PR TITLE
Feature: Add builder methods for `Server`

### PR DESCRIPTION
**Pull request by jgoday**
See  #753.

This PR adds some builder methods to configure zhttp.service.Server:

* withMaxRequestSize(size: Int)
* withPort(port: Int)
* withBind(hostname: String, port: Int)
* withBind(address: InetAddress, port: Int)
* withBind(inetSocketAddress: InetSocketAddress)
* withError[R1](errorHandler: Throwable => ZIO[R1, Nothing, Unit])
* withSsl(sslOptions: ServerSSLOptions)
* withAcceptContinue
* withFlowControl
* withLeakDetectionDisabled
* withSimpleLeakDetection
* withAdvancedLeakDetection
* withParanoidLeakDetection
* withLeakDetection(level: LeakDetectionLevel)
* withKeepAlive
* withConsolidateFlush

1. Added an apply method on Server to create a Server.App from a HttpApp.
2. Added startDefault on Server trait to provide default EventLoopGroup and ServerChannelFactory layers.

It allows the following sample:
```scala
package sample

import zio._
import zhttp.http._
import zhttp.service.Server
import zhttp.service._
import zhttp.service.server._

object Sample extends zio.App {
  def app: HttpApp[Any, Nothing] = Http.collect[Request] {
    case Method.GET -> !! => {
      throw new Exception("???")
      Response.text("bar")
    }
    case Method.POST -> !! => Response.text("foo")
  }

  def server = Server(app)
    .withPort(9090)
    .withError(th => console.putStrLn(th.getMessage()).ignore)
    .withLeakDetection(LeakDetectionLevel.PARANOID)

  override def run(args: List[String]): URIO[zio.ZEnv, ExitCode] =
    server.startDefault.exitCode
}

```

One question about FlowControl:

it seems that by default, flow control is disabled, and when combined with 
current disableFlowControl (adds FlowControl case object).
should not 'FlowControl case object' set 'config flowControl' to true ?
```scala
  // disableFlowControl sets FlowControl case object
  def disableFlowControl: UServer  = Server.FlowControl

  // existence of flowControl case object makes flowControl = false
  private def settings[R1 <: R, E1 >: E](s: Config[R1, E1] = Config()): Config[R1, E1] = self match {
    ...
    case FlowControl          => s.copy(flowControl = false)
  }
```

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 1]
```diff
@@ -36,6 +36,104 @@ sealed trait Server[-R, +E] { self =>
...
+
+  /**
+   * Creates a new server listening on the provided port.
+   */
+  def withPort(port: Int): Server[R, E] = Concat(self, Server.Address(new InetSocketAddress(port)))
+
+  /**
+   * Creates a new server listening on the provided hostname and port.
+   */
+  def withBind(hostname: String, port: Int): Server[R, E] =
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala]:
```suggestion
  def withBinding(hostname: String, port: Int): Server[R, E] =
```

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 2]
```diff
@@ -36,6 +36,104 @@ sealed trait Server[-R, +E] { self =>
...
+    Concat(self, Server.Address(new InetSocketAddress(hostname, port)))
+
+  /**
+   * Creates a new server listening on the provided InetAddress and port.
+   */
+  def withBind(address: InetAddress, port: Int): Server[R, E] =
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala]:
```suggestion
  def withBinding(address: InetAddress, port: Int): Server[R, E] =
```

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 3]
```diff
@@ -36,6 +36,104 @@ sealed trait Server[-R, +E] { self =>
...
+    Concat(self, Server.Address(new InetSocketAddress(address, port)))
+
+  /**
+   * Creates a new server listening on the provided InetSocketAddress.
+   */
+  def withBind(inetSocketAddress: InetSocketAddress): Server[R, E] = Concat(self, Server.Address(inetSocketAddress))
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala]:
```suggestion
  def withBinding(inetSocketAddress: InetSocketAddress): Server[R, E] = Concat(self, Server.Address(inetSocketAddress))
```

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 4]
```diff
@@ -36,6 +36,104 @@ sealed trait Server[-R, +E] { self =>
...
+  /**
+   * Creates a new server using a HttpServerExpectContinueHandler to send a 100 HttpResponse if necessary.
+   */
+  def withAcceptContinue: Server[R, E] = Concat(self, Server.AcceptContinue)
+
+  /**
+   * Creates a new server using netty FlowControlHandler (@see <a
+   * href="https://netty.io/4.1/api/io/netty/handler/flow/FlowControlHandler.html">FlowControlHandler</a>).
+   */
+  def withFlowControl: Server[R, E] = Concat(self, Server.FlowControl)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala]: May be we can use —

```scala
def withFlowControl(enable: Boolean) = ???
```

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 5]
```diff
@@ -36,6 +36,104 @@ sealed trait Server[-R, +E] { self =>
...
+
+  /**
+   * Creates a new server with leak detection disabled (@see <a
+   * href="https://netty.io/4.1/api/io/netty/util/ResourceLeakDetector.Level.html#DISABLED">Disabled Level</a>).
+   */
+  def withLeakDetectionDisabled: Server[R, E] = Concat(self, LeakDetection(LeakDetectionLevel.DISABLED))
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala:99]:
```suggestion
  def withLeakDetection(level: LeakDetectionLevel): Server[R, E] = ???
```

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 6]
```diff
@@ -36,6 +36,104 @@ sealed trait Server[-R, +E] { self =>
...
+   * Creates a new server with paranoid leak detection (@see <a
+   * href="https://netty.io/4.1/api/io/netty/util/ResourceLeakDetector.Level.html#PARANOID">Paranoid Level</a>).
+   */
+  def withParanoidLeakDetection: Server[R, E] = Concat(self, LeakDetection(LeakDetectionLevel.PARANOID))
+
+  /**
+   * Creates a new server with the leak detection level provided (@see <a
+   * href="https://netty.io/4.1/api/io/netty/util/ResourceLeakDetector.Level.html">ResourceLeakDetector.Level</a>).
+   */
+  def withLeakDetection(level: LeakDetectionLevel): Server[R, E] = Concat(self, LeakDetection(level))
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala:101]: I think we should just keep this method, and drop the specific ones. IMO, it is just noise.

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 7]
```diff
@@ -36,6 +36,104 @@ sealed trait Server[-R, +E] { self =>
...
+
+  /**
+   * Creates a new server with netty's HttpServerKeepAliveHandler to close persistent connections when needed (@see <a
+   * href="https://netty.io/4.1/api/io/netty/handler/codec/http/HttpServerKeepAliveHandler.html">HttpServerKeepAliveHandler</a>).
+   */
+  def withKeepAlive: Server[R, E] = Concat(self, KeepAlive)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala]:
```suggestion
  def withKeepAlive(enable: Boolean): Server[R, E] = ???
```

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 8]
```diff
@@ -36,6 +36,104 @@ sealed trait Server[-R, +E] { self =>
...
+
+  /**
+   * Creates a new server with FlushConsolidationHandler to control the flush operations in a more efficient way (@see
+   * <a
+   * href="https://netty.io/4.1/api/io/netty/handler/flush/FlushConsolidationHandler.html">FlushConsolidationHandler<a>).
+   */
+  def withConsolidateFlush: Server[R, E] = Concat(self, ConsolidateFlush)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala]:
```suggestion
  def withConsolidateFlush(enable: Boolean): Server[R, E] = Concat(self, ConsolidateFlush)
```

Comment (jgoday) [reply zio-http/src/main/scala/zhttp/service/Server.scala]: Changed in the last commit

File: zio-http/src/main/scala/zhttp/service/Server.scala (hunk 7 above)
Comment (jgoday) [reply zio-http/src/main/scala/zhttp/service/Server.scala]: Changed in the last commit

File: zio-http/src/main/scala/zhttp/service/Server.scala (hunk 6 above)
Comment (jgoday) [reply zio-http/src/main/scala/zhttp/service/Server.scala:101]: Ok, changed.

File: zio-http/src/main/scala/zhttp/service/Server.scala (hunk 4 above)
Comment (jgoday) [reply zio-http/src/main/scala/zhttp/service/Server.scala]: Changed several methods to use an enable parameter

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 9]
```diff
@@ -38,6 +38,81 @@ sealed trait Server[-R, +E] { self =>
...
+
+  /**
+   * Creates a new server with the following ssl options.
+   */
+  def withSsl(sslOptions: ServerSSLOptions): Server[R, E] = Concat(self, Server.Ssl(sslOptions))
+
+  /**
+   * Creates a new server using a HttpServerExpectContinueHandler to send a 100 HttpResponse if necessary.
+   */
+  def withAcceptContinue: Server[R, E] = Concat(self, Server.AcceptContinue)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala]:
```suggestion
  def withAcceptContinue(enable: Boolean): Server[R, E] = Concat(self, Server.AcceptContinue)
```

Comment (jgoday) [reply zio-http/src/main/scala/zhttp/service/Server.scala]: Changed AcceptContinue

**Comment by tusharmath** [comment]
Thanks @jgoday !

**Review by tusharmath — APPROVED** [review]

**Pull request merged**